### PR TITLE
Migrating to Asciidoc

### LINKED ISSUE #260: Migrate build process + markup to ASCIIdoc(tor)
**As** maintainer, contributor, and user of the RESTful API Guidelines
**I want** the Guideline publishing process to be based on a maintained version of the used software
**So that** the build process is stable and we can make use of latest features and plugins

~~**As** maintainer, contributor, and user of the RESTful API Guidelines~~
~~**I want** the Guideline to be based on the latest stable GitBook version~~
~~**So that** the build process is stable and we can make use of latest features and plugins~~

Related issues: #231, #69, #10, #238.

### PR DESCRIPTION
**Pull request by maxim-tschumak**
This PR implements the migration from Markdown/GitBook to Asciidoc(tor) and includes:

* Translation from Markdown to Asciidoc markup
* New build and deployment scripts
* Introduces RuleId concept and concrete rule ids (and tooling around them)
* Support for the existing links (redirects to the current rule locations)
* Adjusted documentation
* Updated Maintainers list

Please also have a look at the [generated HTML](https://maxim-tschumak.github.io/restful-api-guidelines/).

@tfrauenstein Feel free to award the rule ids [1,99] to the most prominent rules.

Related to #231, #69, #10, #238.
Closes #260.

### CODE REVIEW

**Comment by duergner** [comment]
@maxim-tschumak what's the reason to move over to *RuleId* instead of having speaking URLs?

**Comment by maxim-tschumak** [comment]
@duergner It's a trade-off: having _short easy-to-remember_ RuleIds or _human-readable_ links. As discussed in the Guild, we prefer short identifiers.
It also makes rule identifier reliable (doesn't change when the rule title changes) and allows integrations of another tooling like [Zally](https://github.com/zalando-incubator/zally).

File: MAINTAINERS [hunk 1]
```diff
@@ -1,5 +1,14 @@
+Bill de Hóra <bill.dehora@zalando.ie>
+Michael Duergner <michael.duergner@zalando.fi>
 Paul Ebermann <paul.ebermann@zalando.de>
 Thomas Frauenstein <thomas.frauenstein@zalando.de>
+Christoph Korn <christoph.korn@zalando.de>
+Tronje Krop <tronje.krop@zalando.de>
+Tim Lossen <tim.lossen@zalando.de>
+Andreas Schmidt <andreas.schmidt@zalando.de>
 Willi Schönborn <willi.schoenborn@zalando.de>
+Maxim Tschumak <maxim.tschumak@zalando.de>
```
Comment (ePaul) [inline MAINTAINERS]: Many new maintainers ... Welcome everyone ;-)

File: README.adoc [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+suggestions to the tech community at large.
+
+== Usage
+
+Feel free to use these guidelines as a guidance for your own
+development. Note that we encourage our own teams to use them in order
+to challenge their APIs. As such, you should consider this to be a
+living, evolving document. We will revise and update based on our
+learnings and experiences.
+
```
Comment (ePaul) [inline README.adoc]: Maybe put a link to the BUILD file here?

File: SUMMARY.adoc [hunk 3]
```diff
@@ -0,0 +1,23 @@
+= Summary
+
+* link:chapters/design-principles.adoc[Design Principles]
```
Comment (ePaul) [inline SUMMARY.adoc]: Just for understanding, how does the introduction appear in the output? It is not listed here.

Comment (ePaul) [inline SUMMARY.adoc]: Ah, it comes from index.adoc. Why do we have this file here, then?

Comment (maxim-tschumak) [inline SUMMARY.adoc]: I removed `index.adoc` -> I also see no need for it.

**Comment by ePaul** [comment]
Just a general hint for the next time:
Some files totally change the line breaks, without any content change.
This makes it much harder to review that actually not much changed, as Github's diff view doesn't even notice that they are in fact the same file.

File: chapters/common-data-types.adoc [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+or when doing calculations. Otherwise, you might lose precision.
+Instead, use exact formats like Java’s
+https://docs.oracle.com/javase/8/docs/api/java/math/BigDecimal.html[`BigDecimal`].
+See http://stackoverflow.com/a/3730040/342852[Stack Overflow] for more
+info.
+
+Some JSON parsers (NodeJS’s, for example) convert numbers to floats by
+default. After discussing the
+https://docs.google.com/spreadsheets/d/12wTj-2w39f69XZGwRDrosNc1yWPwQpGgEs_DCt5ODaQ[pros
+and cons], we’ve decided on "decimal" as our amount format. It is not a
```
Comment (ePaul) [inline chapters/common-data-types.adoc]: As this is an internal link, maybe mark is as such?

File: chapters/compatibility.adoc [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+* Support redirection in case an URL has to change
+(https://en.wikipedia.org/wiki/HTTP_301[301 Moved Permanently]).
+
+[#108]
+== {MUST} Prepare Clients To Not Crash On Compatible API Extensions
+
+Service clients should apply the robustness principle:
+
+* Be conservative with API requests and data passed as input, e.g. avoid
+to exploit definition eficits like passing megabytes for strings with
```
Comment (ePaul) [inline chapters/compatibility.adoc]: eficits → deficits.
(I know, this didn't change really with this PR, but as everything is marked as changed anyways ...)

File: chapters/compatibility.adoc [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+
+In specific situations, where a (known) input field is not needed
+anymore, it either can stay in the API definition with "not used
+anymore" description or can be removed from the API definition as long
+as the server ignores this specific parameter.
+
+[#110]
+== {MUST} Always Return JSON Objects As Top-Level Data Structures To Support Extensibility
+
+In a response body, you must always return a JSON objects (and not e.g.
```
Comment (ePaul) [inline chapters/compatibility.adoc]: objects → object

File: chapters/data-formats.adoc [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+payload must follow https://tools.ietf.org/html/rfc7159[RFC-7159] by
+having (if possible) a serialized object as the top-level structure,
+since it would allow for future extension. This also applies for
+collection resources where one naturally would assume an array. See <<160>> for
+an example.
+
+[#168]
+== {MAY} Use non JSON Media Types for Binary Data or Alternative Content Representations
+
+Other media types may be used in following cases: * Transferring binary
```
Comment (ePaul) [inline chapters/data-formats.adoc]: Here some line breaks (and therefore the list structure) got lost.

File: chapters/data-formats.adoc [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+[#169]
+== {MUST} Use Standard Date and Time Formats
+
+=== JSON Payload
+
+Read more about date and time format in <<125>>.
+
+=== HTTP headers
+
+Http headers including the proprietary headers. Use the
```
Comment (ePaul) [inline chapters/data-formats.adoc]: Not new, but this is not a sentence. Can we get rid of the `.` and just join the two sentences here?

File: chapters/data-formats.adoc [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+Whenever an API defines a property of type `number` or `integer`, the
+precision must be defined by the format as follows to prevent clients
+from guessing the precision incorrectly, and thereby changing the value
+unintentionally:
+
+[cols="15%,15%,70%",options="header",]
+|=====================================================================
+|type |format |specified value range
+|integer |int32 |integer between -231 and 231-1
+|integer |int64 |integer between -263 and 263-1
```
Comment (ePaul) [inline chapters/data-formats.adoc]: Those ranges are a bit small ;-)

The exponents got lost here.

File: chapters/design-principles.adoc [hunk 10]
```diff
@@ -0,0 +1,94 @@
+[[principles]]
+= Principles
+
+[[api-as-a-product]]
+== API as a Product
+
+As mentioned above, Zalando is transforming from an online shop into an
+expansive fashion platform comprising a rich set of products following a
+Software as a Platform (SaaS) model for our business partners. As a
```
Comment (ePaul) [inline chapters/design-principles.adoc]: Hmm, "Software as a Platform" doesn't have the abbreviation "SaaS" (that is software as a service).

@tfrauenstein Is the abbreviation or the full name wrong? (It was added in your PR #159.)

File: chapters/design-principles.adoc [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+components, whether they provide functionality via the Internet or via
+the intranet as larger application elements. We strive to build
+interoperating distributed systems that different teams can evolve in
+parallel.
+
+An important principle for (RESTful) API design and usage is Postel's
+Law, aka http://en.wikipedia.org/wiki/Robustness_principle[the
+Robustness Principle] (RFC 1122):
+
+* Be liberal in what you accept, be conservative in what you send
```
Comment (ePaul) [inline chapters/design-principles.adoc]: Hmm, not sure if that is still applicable, as we derive from it in the Compatibility chapter, specifically rule 109.
(Also not new.)

File: chapters/events.adoc [hunk 12]
```diff
@@ -0,0 +1,795 @@
...
+* Events that are the result of a computation, such as a matching
+algorithm, or the generation of enriched data, and which might not be
+stored as entity by the service.
+
+[#206]
+== {MUST} Permissions on events must correspond to API permissions
+
+If a resource can be read synchronously via a REST API and read
+asynchronously via an event, the same read-permission must apply: We
+want to protect access to data, not the way data is accessed.
```
Comment (ePaul) [inline chapters/events.adoc]: I guess we should reconsider this (not now, sometime later), now that Nakadi has a new permission model, and got rid of the permissions.

File: chapters/http.adoc [hunk 13]
```diff
@@ -0,0 +1,486 @@
...
+* GET requests for individual resources will usually generate a 404 if
+the resource does not exist
+* GET requests for collection resources may return either 200 (if the
+listing is empty) or 404 (if the list is missing)
+* GET requests must NOT have request body payload
+
+*Note:* GET requests on collection resources should provide a sufficient
+filter mechanism as well as <<pagination>>.
+
+=== GET with Body
```
Comment (ePaul) [inline chapters/http.adoc]: (Not new)
I think we should put the title in quotes ("GET with Body"), as there is no such thing in HTTP.

File: chapters/http.adoc [hunk 14]
```diff
@@ -0,0 +1,486 @@
...
+errors. In most cases it's not useful to document technical errors that
+are not in control of the service provider unless the status code convey
+application-specific semantics. The list of status code that can be
+omitted from API specifications includes but is not limited to: -
+`401 Unauthorized` - `403 Forbidden` - `404 Not Found` unless it has
+some additional semantics - `405 Method Not Allowed` -
+`406 Not Acceptable` - `408 Request Timeout` - `413 Payload Too Large` -
+`414 URI Too Long` - `415 Unsupported Media Type` -
+`500 Internal Server Error` - `502 Bad Gateway` -
+`503 Service Unavailable` - `504 Gateway Timeout`
```
Comment (ePaul) [inline chapters/http.adoc]: This list got broken. Please reformat it.

**Comment by ePaul** [comment]
**A note about my comments:** most don't need to be fixed right now (some do, which are about formatting issues introduced by this change). But as this marks about everything as changed, I used the chance to review everything (not totally finished yet, but I've to go home now).

Feel free to fix just the formatting issues, I can open separate pull requests afterwards (or issues where there is still some need to discuss).

File: chapters/design-principles.adoc (hunk 10 above)
Comment (maxim-tschumak) [inline chapters/design-principles.adoc]: I've corrected it to `SaaP`.

File: chapters/design-principles.adoc (hunk 11 above)
Comment (maxim-tschumak) [inline chapters/design-principles.adoc]: Should I add a reference to `109`?

Btw, @ePaul, you are the first person who referenced to a rule by its new ID :blush:

**Comment by maxim-tschumak** [comment]
@ePaul Thanks for the review! - I've fixed the layout problems and the minor content improvements. (and also updated the [generated HTML](https://maxim-tschumak.github.io/restful-api-guidelines/))

@all Please add the minor, nice-to-have improvements to #283.
Reason: In order to continue working on the content of the Guidelines, we have to merge this PR as soon as possible (to avoid merge conflicts).

File: chapters/hyper-media.adoc [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+real value in our SOA context where client and server interact via REST
+APIs and provide complex business functions as part of our e-commerce
+SaaS platform.
+
+Our major concerns regarding the promised advantages of HATEOAS (see
+also
+https://www.infoq.com/news/2014/03/rest-at-odds-with-web-apis%20for%20detailed%20discussion[RESTistential
+Crisis over Hypermedia APIs],
+https://jeffknupp.com/blog/2014/06/03/why-i-hate-hateoas/[Why I Hate
+HATEOAS] and others): - We follow the <<100,API First principle>>
```
Comment (ePaul) [inline chapters/hyper-media.adoc]: Another broken list here.

File: chapters/json-guidelines.adoc [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+These guidelines provides recommendations for defining JSON data at
+Zalando. JSON here refers to
+http://www.rfc-editor.org/rfc/rfc7159.txt[RFC 7159] (which updates
+https://www.ietf.org/rfc/rfc4627.txt[RFC 4627]), the “application/json”
+media type and custom JSON media types defined for APIs. The guidelines
+clarifies some specific cases to allow Zalando JSON data to have an
+idiomatic form across teams and services.
+
+[#117]
+== {MUST} Use Consistent Property Names
```
Comment (ePaul) [inline chapters/json-guidelines.adoc]: This was previously a level-2 title above the following level-3 rules. Now it looks like it is at the same level as the rules. (Same with rule 121 below.)

File: chapters/json-guidelines.adoc [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+as required). However that doesn't prevent clients and servers sending
+and receiving those fields with null values. Also, in some cases null
+may be a meaningful value - for example, JSON Merge Patch
+https://tools.ietf.org/html/rfc7386[RFC 7382]) using null to indicate
+property deletion.
+
+[#124]
+== {SHOULD} Empty array values should not be null
+
+Empty array values can unambiguously be represented as the the empty
```
Comment (ePaul) [inline chapters/json-guidelines.adoc]: Duplicated "the".

File: chapters/pagination.adoc [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+API clients
+* Use case: Jump to a certain page
+** If jumping to a particular page in a range (e.g., 51 of 100) is
+really a required use case, cursor-based navigation is not feasible
+* Variability of data may lead to anomalies in result pages
+** Offset-based pagination may create duplicates or lead to missing
+entries if rows are inserted or deleted between two subsequent paging
+requests.
+** When using cursor-based pagination, paging cannot continue when the
+cursor entry has been deleted while fetching two pages
```
Comment (ePaul) [inline chapters/pagination.adoc]: Not new, but I highly doubt this argument. (I guess it depends on the kind of cursor you are using.)

File: chapters/performance.adoc [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+APIs should support techniques for reducing bandwidth based on client
+needs. This holds for APIs that (might) have high payloads and/or are
+used in high-traffic scenarios like the public Internet and
+telecommunication networks. Typical examples are APIs used by mobile web
+app clients with (often) less bandwidth connectivity. (Zalando is a
+'Mobile First' company, so be mindful of this point.)
+
+Common techniques include:
+
+* gzip compression
```
Comment (ePaul) [inline chapters/performance.adoc]: this could be a link to <<156>>.

File: chapters/performance.adoc [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+* querying field filters to retrieve a subset of resource attributes
+(see <<156>> below)
```
Comment (ePaul) [inline chapters/performance.adoc]: s/156/157/

File: chapters/proprietary-headers.adoc [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+
+`X-` headers were initially reserved for unstandardized parameters, but
+the usage of `X-` headers is deprecated
+(https://tools.ietf.org/html/rfc6648[RFC-6648]). This complicates the
+contract definition between consumer and producer of an API following
+these guidelines, since there is no aligned way of using those headers.
+Because of this, the guidelines restrict which `X-` headers can be used
+and how they are used.
+
+The Internet Engineering Task Force's states in
```
Comment (ePaul) [inline chapters/proprietary-headers.adoc]: Not new, but the `'s` here is wrong. (Actually I would reword the whole paragraph, will do in a separate PR.)

File: chapters/proprietary-headers.adoc [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+https://tools.ietf.org/html/rfc6648[RFC-6648] that company specific
+header' names should incorporate the organization's name. We aim for
+backward compatibility, and therefore keep the `X-` prefix.
+
+The following proprietary headers have been specified by this guideline
+for usage so far. Remember that HTTP header field names are not
+case-sensitive.
+
+[cols="15%,10%,60%,15%",options="header",]
```
Comment (ePaul) [inline chapters/proprietary-headers.adoc]: Do we really need to define the column widths manually? Isn't this something a browser can figure out based on the table content?

File: chapters/resources.adoc [hunk 23]
```diff
@@ -0,0 +1,196 @@
...
+== {SHOULD} Only Use UUIDs If Necessary
+
+Generating IDs can be a scaling problem in high frequency and near real
+time use cases. UUIDs solve this problem, as they can be generated
+without collisions in a distributed, non-coordinated way and without
+additional server roundtrips.
+
+However, they also come with some disadvantages:
+
+* pure technical key without meaning; not ready for naming or name scope
```
Comment (ePaul) [inline chapters/resources.adoc]: (Not new.) This list looks like each sentence there should be start with "they are ...". Maybe put this at the end of the paragraph above?

File: chapters/resources.adoc [hunk 24]
```diff
@@ -0,0 +1,196 @@
...
+for instance:
+
+[source,http]
+----
+/carts/1681e6b88ec1/cart-items/1
+----
+
+However, if the resource can be accessed directly via its unique id,
+then the API should expose it as a top-level resource. For example,
+customer is a collection for sales orders; however, sales orders have
```
Comment (ePaul) [inline chapters/resources.adoc]: is → has

File: chapters/json-guidelines.adoc (hunk 16 above)
Comment (maxim-tschumak) [inline chapters/json-guidelines.adoc]: This is done on purpose - the idea is to have all the rules on the same level.

File: chapters/proprietary-headers.adoc (hunk 22 above)
Comment (maxim-tschumak) [inline chapters/proprietary-headers.adoc]: Unfortunately, no. The browser renders all the columns with the same width (when using the default CSS).

File: chapters/performance.adoc (hunk 20 above)
Comment (tkrop) [inline chapters/performance.adoc]: @maxim-tschumak Have you checked all referenced up from the missing one? Which was it? `116`?

File: chapters/proprietary-headers.adoc (hunk 22 above)
Comment (tkrop) [inline chapters/proprietary-headers.adoc]: @maxim-tschumak Have you checked that the style uses `table-layout: auto`? This should provide what @ePaul  expects.

**Comment by wssbck** [comment]
Looks great. One more addition perhaps - it would be very cool if headers in the navigation on the left had some highlighting for currently browsed section.

**Comment by tkrop** [comment]
@wssbck 😸 we have this on the todo-list #283.

**Comment by tfrauenstein** [comment]
👍  look great -- thx.

**Comment by tkrop** [comment]
👍

**Comment by ePaul** [comment]
@tfrauenstein do you want to change numbers before we merge it, or do we risk breaking links if you do it later?

**Comment by ePaul** [comment]
:+1:

**Comment by tkrop** [comment]
:+1:

**Comment by maxim-tschumak** [comment]
:+1:

**Pull request merged**